rust/main/chains/hyperlane-starknet/src/mailbox.rs

```diff
@@ -0,0 +1,258 @@
+                .contract
+                .nonce()
+                .call()
+                .await
+                .map_err(Into::<HyperlaneStarknetError>::into)?,
+        };
+
+        Ok(nonce)
+    }
+
+    #[instrument(skip(self))]
+    async fn delivered(&self, id: H256) -> ChainResult<bool> {
+        let bytes = id.as_bytes();
+        let (high_bytes, low_bytes) = bytes.split_at(16);
+        let high = BigEndian::read_u128(high_bytes);
+        let low = BigEndian::read_u128(low_bytes);
+        Ok(self
+            .contract
+            .delivered(&StarknetU256 { low, high })
+            .call()
+            .await
+            .map_err(Into::<HyperlaneStarknetError>::into)?)
+    }
+
+    #[instrument(skip(self))]
+    async fn default_ism(&self) -> ChainResult<H256> {
+        let address = self
+            .contract
+            .get_default_ism()
+            .call()
+            .await
+            .map_err(Into::<HyperlaneStarknetError>::into)?;
+        Ok(address.0.into())
+    }
+
+    #[instrument(skip(self))]
+    async fn recipient_ism(&self, recipient: H256) -> ChainResult<H256> {
+        let address = self
+            .contract
+            .recipient_ism(&StarknetU256::from_bytes_be(&recipient.to_fixed_bytes()))
+            .call()
+            .await
+            .map_err(Into::<HyperlaneStarknetError>::into)?;
+        Ok(address.0.into())
+    }
+
+    #[instrument(skip(self), fields(metadata=%bytes_to_hex(metadata)))]
+    async fn process(
+        &self,
+        message: &HyperlaneMessage,
+        metadata: &[u8],
+        tx_gas_limit: Option<U256>,
+    ) -> ChainResult<TxOutcome> {
+        let contract_call = self
+            .process_contract_call(message, metadata, tx_gas_limit)
+            .await?;
+        let tx = contract_call
+            .send()
+            .await
+            .map_err(|e| HyperlaneStarknetError::AccountError(e.to_string()))?;
+        let invoke_tx_receipt =
+            get_transaction_receipt(&self.provider.rpc_client(), tx.transaction_hash).await;
+        match invoke_tx_receipt {
+            Ok(MaybePendingTransactionReceipt::Receipt(TransactionReceipt::Invoke(receipt))) => {
+                return Ok(receipt.try_into()?);
+            }
+            Ok(MaybePendingTransactionReceipt::PendingReceipt(
+                PendingTransactionReceipt::Invoke(receipt),
+            )) => {
+                return Ok(receipt.try_into()?);
+            }
+            _ => {
+                return Err(HyperlaneStarknetError::InvalidTransactionReceipt.into());
+            }
+        }
+    }
+
+    #[instrument(skip(self), fields(msg=%message, metadata=%bytes_to_hex(metadata)))]
+    async fn process_estimate_costs(
+        &self,
+        message: &HyperlaneMessage,
+        metadata: &[u8],
+    ) -> ChainResult<TxCostEstimate> {
+        let _contract_call = self.process_contract_call(message, metadata, None).await?;
+
+        Ok(TxCostEstimate::default())
```

Comment:
How would one actually estimate the gas cost of a call? Is there some RPC / provider method for this?

Reply:
done